engine/src/main/java/org/hibernate/validator/internal/engine/valueextraction/ValueExtractorManager.java

```diff
@@ -351,102 +348,82 @@ private static boolean isClassPresent(String className, boolean fallbackOnTCCL)
 		}
 	}
 
-	private final class ValueExtractorCache {
+	private final class ValueExtractorResolutionCache {
 
-		private final ConcurrentHashMap<ValueExtractorCacheKey, Set<ValueExtractorDescriptor>> valueExtractorCache;
-		private final Set<ValueExtractorCacheKey> nonContainerTypes;
+		private final ConcurrentHashMap<ValueExtractorCacheKey, Set<ValueExtractorDescriptor>> possibleValueExtractorsByRuntimeTypes;
+		private final Set<Class<?>> nonContainerTypes;
 
-		public ValueExtractorCache() {
-			this.valueExtractorCache = new ConcurrentHashMap<>();
+		public ValueExtractorResolutionCache() {
+			this.possibleValueExtractorsByRuntimeTypes = new ConcurrentHashMap<>();
 			this.nonContainerTypes = new HashSet<>();
 		}
 
-		public Set<ValueExtractorDescriptor> getValueExtractors( Set<ValueExtractorDescriptor> valueExtractorCandidates,
-							Class<?> valueType ) {
-			ValueExtractorCacheKey cacheKey = new ValueExtractorCacheKey( valueType, /*something here*/ null );
-
-			if ( nonContainerTypes.contains( cacheKey ) ) {
+		public Set<ValueExtractorDescriptor> getValueExtractors(Type declaredType, Class<?> runtimeType, TypeVariable<?> typeParameter ) {
+			if ( nonContainerTypes.contains( runtimeType ) ) {
 				return Collections.emptySet();
 			}
+			ValueExtractorCacheKey cacheKey = new ValueExtractorCacheKey( runtimeType, typeParameter );
 
-			Set<ValueExtractorDescriptor> valueExtractorDescriptors = valueExtractorCache.get( cacheKey );
+			Set<ValueExtractorDescriptor> valueExtractorDescriptors = possibleValueExtractorsByRuntimeTypes.get( cacheKey );
 			if ( valueExtractorDescriptors == null ) {
-				valueExtractorDescriptors = getMaximallySpecificValueExtractors(
-						valueType,
-						valueExtractorCandidates
-				);
+				Set<ValueExtractorDescriptor> possibleValueExtractors = getMaximallySpecificValueExtractors( runtimeType );
```

Comment:
I'm not sure you got the logic right here.

We should select the type compliant VEs before filtering to get the maximally specific ones. If you filter the maximally specific VEs first, you might miss a VE that is not maximally specific in the general case BUT is in fact the maximally specific one that is container element compliant.

You cannot use `getMaximallySpecificValueExtractors(Class<?> valueType)` as it should only be used when we have an automatic unwrapping. We can also use it in the case the `@Valid` annotation is on the whole container as in this case we don't have to be container element compliant but definitely not when we are looking for a container element compliant one.

See the resolution algorithm in detail here: http://beanvalidation.org/2.0/spec/#constraintdeclarationvalidationprocess-validationroutine-valueextractorresolution-algorithm-cascaded .

I'm not sure it's easy to build a test case to make what you did fail but I think it would be worth it to at least try.

If I'm not clear, don't hesitate to ping me, it's not an easy subject...